### ISSUE TITLE
Should `ApolloClient.readQuery` have it's options defined as `DataProxy.readQueryOptions`?

### ISSUE THREAD

**Issue by baseten** [apollographql/apollo-client#11248]
The implementation of `ApolloClient.readQuery` defines its `option` like so:

https://github.com/apollographql/apollo-client/blob/main/src/core/ApolloClient.ts#L415-L420

However the underlying `DataProxy` defines `ReadQueryOptions` like so:

https://github.com/apollographql/apollo-client/blob/main/src/cache/core/types/DataProxy.ts#L58-L77

Is this done intentionally to prevent `ApolloClient.readQuery` from receiving the full cache options? Currently (although not type-safe), it's possible to pass `returnPartialData: true` to `client.readQuery`. This would be super useful for our use case, but we don't want to use it if it's an unintentional escape hatch.

**Comment by jerelmiller** [comment]
Hey @baseten 👋 

This appears to be more of an oversight rather than intentional restriction. Since `client.readQuery` just forwards the call to `client.cache.readQuery`, there is no reason those types shouldn't line up. Updating to `DataProxy.ReadQueryOptions` is a backwards compatible change, so this should be safe to change.

Would you be willing to open a PR to update the type here? We'd be happy to accept it.

One note on a small tweak I'd make when doing so. Notice how `DataProxy.ReadQueryOptions` accepts an [`optimistic` option](https://github.com/apollographql/apollo-client/blob/bc055e0683e87b9445e321f73857f4a91b20a9ce/src/cache/core/types/DataProxy.ts#L70), yet our function call has an `optimistic` argument as the 2nd argument. For backwards compatibility, we'll want to leave this alone, but to avoid confusion I think we should omit `optimistic` from `options` in the type since the `optimistic` argument is going to [override](https://github.com/apollographql/apollo-client/blob/bc055e0683e87b9445e321f73857f4a91b20a9ce/src/cache/core/cache.ts#L141) anything in `options`.

In other words, the type should be:

```ts
public readQuery<T = any, TVariables = OperationVariables>(
  options: Omit<DataProxy.ReadQueryOptions<TVariables, T>, 'optimistic'>,
  optimistic: boolean = false
): T | null {
  return this.cache.readQuery<T, TVariables>(options, optimistic);
}
```

Let me know if you'd be willing to open a PR!

**Comment by jerelmiller** [comment]
Hey @baseten 👋 

As of Apollo Client 4.0, the `DataProxy` namespace has been removed. We've reworked our types quite a bit to make them more discoverable by colocating many of these options/result types along with the API that owns them. In this particular case, 4.0 now exposes these options as [`ApolloClient.ReadQueryOptions`](https://github.com/apollographql/apollo-client/blob/60fad13d2695ea3ccd5d8323c2de52e37201f4fb/src/core/ApolloClient.ts#L1050). These types have been flattened out quite a bit to avoid some of the `extends Pick`/`Omit` mess from some of the types in v3. That [`returnPartialData` option](https://github.com/apollographql/apollo-client/blob/60fad13d2695ea3ccd5d8323c2de52e37201f4fb/src/core/ApolloClient.ts#L412) is now properly included in that type.

I'd highly recommend upgrading to 4.0 if you can. Please see our [migration guide](https://www.apollographql.com/docs/react/migrating/apollo-client-4-migration) for instructions on how to migrate when you're ready to do so 🙂. Since this should be addressed in 4.0, I'm going to go ahead and close this issue as completed. If you see any issues after upgrading, please open a new issue and we'd be happy to take a look. Thanks!

**Comment by baseten** [comment]
Hey @jerelmiller 👋 

Thanks for the message and the update! 🙏
